CODING_CONVENTIONS.md

```diff
@@ -0,0 +1,145 @@
+  - For other annotations, doesn't matter; let your IDE decide
+
+## Naming
+* Classes named in CapitalCamelCase
+* Fields and methods named in camelCase
+  - Final static fields may be named in SCREAMING_SNAKE_CASE
+* Localized messages should be named like m_message_name
+  - And that is the only place where snake_case is acceptable
+* Use prefixes where their use has been already estabilished (such as ExprSomeRandomThing)
+  - Otherwise, use postfixes (such as LoopSection)
+  
+## Comments
+* Prefer to comment *why* you're doing things instead of how you're doing them
+  - In case the code is particularly complex, though, do both!
+* Write all comments in readable English
+  - Remember the basic grammar rules
+  - Fun is allowed, if that does not hurt the readability
+* No matter if a comment contains a sentence or not, capitalize the first letter
+* For single line comments
+  - Always start with a space; it improves readability
+  - You *may* put them at ends of lines, if the line doesn't get too wide due to that
+
+Your comments should look something like these:
+```java
+// A single line comment
+
+/**
+ * This is a Javadoc comment.
+ * @param first Description of the first parameter.
+ * @param second Description of the second parameter.
+ * @return Description of the return value.
+ */
+ 
+/*
+ * This is a block comment.
+ */
+```
+
+## Language Features
+* Java 8 source and binary compatibility
+  - JDK 8 must be enough to develop Skript
+  - Users must not need JRE newer than version 8
+* Java 9 and 10 must also work
+  - Absolutely no dirty deep reflection hacks
+* It is recommended to make fields final, if they are effectively final
+  - Performance impact is be minimal, but it makes code cleaner
+* Local variables and method parameters should not be declared final
+* Methods should be declared final only rarely
+  - Static methods should not *ever* be declared final
+* Use @Override where applicable
+
+## Nullness
+* All fields, method parameters and their return values are non-null by default
+  - Exceptions: Github API JSON mappings, Metrics
+* When something is nullable, mark it as so
+* Only ignore nullness errors when a variable is effectively non-null, but compiler doesn't recognize it
+  - Most common example is syntax elements, which are not initialized using a constructor
+* Use assertions liberally: if you're sure something is not null, assert so to the compiler
+  - Makes finding bugs easier for developers
+* Avoid checking non-null values for nullability
+  - Unless working around buggy addons, it is rarely necessary
+  - This is why ignoring nullness errors is particularly dangerous
+
+## Portability
+* **Absolutely no NMS-reliant code**
+  - No matter *how* it is done, it belongs to addons
+  - Updating for new Minecraft versions is already painful enough
+  ```java
+  import net.minecraft.server.v1_13_1.*; // BAD
+  ```
+* Target Minecraft versions are 1.9, 1.10, 1.11, 1.12 and 1.13 (including all revisions)
```

Comment:
I think this goes against what we stated before about the supported MC versions in the README. We officially don't support revisions.

Reply:
Yeah, it does. I'm not sure what I thought when I wrote this one, because even at that point I was *not* testing with all revisions.